### PR TITLE
Fix serving files with diacritics in name under their original name properly

### LINKED ISSUE #1776: Cherrypy doesn`t encode UTF-8 filename in HTTP header "Content-Disposition" and returns "download" browser default filename instead.
**I'm submitting a ...**
- [X] bug report
- [ ] feature request
- [ ] question about the decisions made in the repository

**Do you want to request a *feature* or report a *bug*?**
Report a bug.

**What is the current behavior?**
cherrypy.lib.static.serve_file() method sets wrong "Content-Disposition" and doesn't encode UTF-8 filename properly. It causes error with browser suggested filename - when downloading the file, the value is "download" and it doesn't matter what the original filename is. This could be fixed with RFC6266 (https://tools.ietf.org/html/rfc6266#appendix-D) solution, e.g. filename="my-resume.pdf"; filename*=utf-8''My%20R%C3%A9sum%C3%A9.pdf.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a screenshots and logs of the problem. If you can, show us your code.**
Code:
```
return cherrypy.lib.static.serve_file(
    path='file_path',
    content_type='file_mime_type',
    disposition='attachment',
    name='has_utf-8_character_☃.html',
    debug=True)
```
DEBUG:
TOOLS.STATIC Content-Disposition: 'attachment; filename="has_utf-8_character_☃.html"'

**What is the expected behavior?**
DEBUG:
TOOLS.STATIC Content-Disposition: 'attachment; filename="has_utf-8_character_.html"; filename*=utf-8''has_utf-8_character_%E2%98%83.html'

**What is the motivation / use case for changing the behavior?**
To be able return file with czech diacritics in name for download under its original name.

**Please tell us about your environment:**

- Cheroot version: v6.5.4
- CherryPy version: v18.0.1
- Python version: v3.6.6
- OS: Windows 10
- Browser: [all]

**Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, e.g. stackoverflow, gitter, etc.)
https://stackoverflow.com/questions/1361604/how-to-encode-utf8-filename-for-http-headers-python-django/8996249#8996249

### PR DESCRIPTION
**Pull request by jrxpert**
**What kind of change does this PR introduce?**
  - [x] bug fix
  - [ ] feature
  - [ ] docs update
  - [ ] tests/coverage improvement
  - [ ] refactoring
  - [ ] other

**What is the related issue number (starting with `#`)**
Fixes #1776

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/cherrypy/cherrypy/issues/1776

**What is the new behavior (if this is a feature change)?**

**Other information**:

**Checklist**:

  - [x] I think the code is well written
  - [x] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [x] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [ ] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

### CODE REVIEW

**Comment by webknjaz** [comment]
Doesn't #1695 address this?

**Comment by jrxpert** [comment]
Hello, thank you for response. Pull request you linked solves diacritics for file upload, this pull request solves diacritics for file download.

File: cherrypy/lib/static.py [hunk 1]
```diff
@@ -28,6 +29,21 @@ def _setup_mimetypes():
...
 
 
+def _utf8_content_disposition(disposition, file_name):
+    """Create HTTP header for downloading a file with UTF-8 filename.
+
+    See this and related answers:
+    https://stackoverflow.com/a/8996249/2173868
+    """
+    ascii_name = unicodedata.normalize(
+        'NFKD', file_name).encode('ascii', 'ignore').decode()
```
Comment (webknjaz) [inline cherrypy/lib/static.py]: I'm not in favor of using `ignore` for error processing. I think it shadows errors. What is the reason behind using it?

File: cherrypy/lib/static.py [hunk 2]
```diff
@@ -28,6 +29,21 @@ def _setup_mimetypes():
...
+    header = '{}; filename="{}"'.format(disposition, ascii_name)
+    if ascii_name != file_name:
```
Comment (webknjaz) [inline cherrypy/lib/static.py]: Should we make this conditional? Adding it unconditionally would simplify the code.

File: cherrypy/lib/static.py [hunk 3]
```diff
@@ -28,6 +29,21 @@ def _setup_mimetypes():
...
 
+def _utf8_content_disposition(disposition, file_name):
+    """Create HTTP header for downloading a file with UTF-8 filename.
+
+    See this and related answers:
+    https://stackoverflow.com/a/8996249/2173868
+    """
+    ascii_name = unicodedata.normalize(
+        'NFKD', file_name).encode('ascii', 'ignore').decode()
+    header = '{}; filename="{}"'.format(disposition, ascii_name)
```
Comment (webknjaz) [inline cherrypy/lib/static.py]: Let's use a Unicode literal here to be explicit:
```suggestion
    header = u'{}; filename="{}"'.format(disposition, ascii_name)
```

File: cherrypy/lib/static.py [hunk 4]
```diff
@@ -28,6 +29,21 @@ def _setup_mimetypes():
 _setup_mimetypes()
 
 
+def _utf8_content_disposition(disposition, file_name):
+    """Create HTTP header for downloading a file with UTF-8 filename.
+
+    See this and related answers:
+    https://stackoverflow.com/a/8996249/2173868
+    """
+    ascii_name = unicodedata.normalize(
```
Comment (webknjaz) [inline cherrypy/lib/static.py]: Could you please change the style here to
```python
var_name = (
    multi().line().
    chain().of().calls()
)
```
?

File: cherrypy/lib/static.py [hunk 5]
```diff
@@ -28,6 +29,21 @@ def _setup_mimetypes():
...
+        'NFKD', file_name).encode('ascii', 'ignore').decode()
+    header = '{}; filename="{}"'.format(disposition, ascii_name)
+    if ascii_name != file_name:
+        quoted_name = urllib.parse.quote(file_name)
+        header += '; filename*=UTF-8\'\'{}'.format(quoted_name)
```
Comment (webknjaz) [inline cherrypy/lib/static.py]:
```suggestion
        header += u'; filename*=UTF-8\'\'{}'.format(quoted_name)
```

**Review by webknjaz — APPROVED** [review]
I like this! Please edit it a bit and I'll hit merge :)

**Comment by webknjaz** [comment]
@jrxpert

A few more notes:
1) Commit message `Fix serving files with diacritics in name under their original name properly` doesn't fit 50 chars limit, _could you please change it to_:

    ```
    Fix serving files with diacritics in name under their original name

    This implements adding extra ``filename*`` parameter in Content-disposition
    header when serving content with non-ASCII filenames. As per RFC 6266
    recommendations in the Appendix D.

    Refs:
    * https://stackoverflow.com/a/8996249/2173868
    * https://tools.ietf.org/html/rfc6266#appendix-D

    Fixes #1776
    ```
2) I noted that your commit isn't linked you your GitHub account. That's probably because your local Git configuration contains email which doesn't match the one you set up in GitHub. You may want to fix that :)

**Comment by webknjaz** [comment]
@jrxpert should I expect updates here anytime soon?

File: cherrypy/lib/static.py (hunk 1 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Try use it with default "errors" setting (strict), you will get "UnicodeEncodeError: 'ascii' codec can't encode character '\u2603' in position 20: ordinal not in range(128)" when running tests. Thats protect it against 'ascii' codec unknown UTF8 character error.

File: cherrypy/lib/static.py (hunk 2 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: We don`t have to, its ok to simplify it.

File: cherrypy/lib/static.py (hunk 3 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Yes, good idea.

File: cherrypy/lib/static.py (hunk 4 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Yes, no problem.

**Comment by jrxpert** [comment]
I revert previous commit and done it again with few changes related to comments. Because of problems with local `pre-commit` installation, i did this commit without that. Tests are working.

File: cherrypy/lib/static.py (hunk 2 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: We don`t have to, its ok to simplify it.

File: cherrypy/lib/static.py (hunk 3 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Yes, good idea.

File: cherrypy/lib/static.py (hunk 4 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Yes, no problem.

File: cherrypy/lib/static.py (hunk 1 above)
Comment (jrxpert) [reply cherrypy/lib/static.py]: Try use it with default "errors" setting (strict), you will get "UnicodeEncodeError: 'ascii' codec can't encode character '\u2603' in position 20: ordinal not in range(128)" when running tests. Thats protect it against 'ascii' codec unknown UTF8 character error.

**Pull request closed without merging**